### PR TITLE
switch out protobuf dependency for gpb

### LINKED ISSUE #4: prefer gpb to elixir-protobuf/protobuf
According to hex.pm, the status of protobuf projects is roughly

| package | recent downloads | notes |
|--------|---------------|---------|
| [`tomas-abrahamsson/gpb`](https://github.com/tomas-abrahamsson/gpb) | 436_657 | primarily targets erlang fwict |
| [`elixir-protobuf/protobuf`](https://github.com/elixir-protobuf/protobuf) | 223_097 | name collision with below |
| [`bitwalker/exprotobuf`](https://github.com/bitwalker/exprotobuf) | 58_526 | name collision with above |
| [`ahamez/protox`](https://github.com/ahamez/protox) | 4_627 | requires `protoc` installed on compilation machine |

Currently spear uses `elixir-protobuf/protobuf`, but this causes incompatibility with any project that uses bitwalker's implementation, including [`exponentially/extreme`](https://github.com/exponentially/extreme). For use in a service/application, I wouldn't mind just choosing either but since this is a library we need to be as dependency-permissive as possible.

So because of the module conflicts, I consider both `elixir-protobuf/protobuf` and `bitwalker/exprotobuf` non-starters.

Similarly I'm not comfortable adding installation instructions for `protoc` to this library (required for compiling `protox` definitions). Acceptable in a service/application but this library should bite the bullet of (com/trans)piling `.proto` files into an elixir/erlang format.

Ideally I'd like to write two separate libraries: a parser/transpiler for proto into elixir/erlang terms and then a protobuf engine which performs the encoding/decoding. That way libraries can use the transpiler to generate some files and then only export a dependency on the engine.

However, I only have so much time and energy I realize that with my strong opinions about what is right and wrong library behavior I am poking the library version of this XKCD bear:

`:gpb` seems to be the only good option currently. It isn't completely ideal (~e.g. it fails to install with `Mix.install/1`~ :facepalm: I have a PR open in gpb to fix this) but I might need to just pick my poison and move on.

### PR DESCRIPTION
**Pull request by the-mikedavis**
closes #4 

that issue has all the reasons laid out of why we want to do this

can't really merge this PR until my PR to gpb gets merged and a new release pushed, or we can add a line to the actions CI to do the workaround (`mkdir -p deps/gpb/ebin` should do it)

---

`:gpb` (an erlang library) takes a _very_ erlangy approach to protobufs:

- the code generator (`.proto` -> `.erl`+`.hrl`) is written as an escript
- the default output is erlang records

---

### Records

Records are actually super cool, albeit a bit hard to use in Elixir because there aren't out-of-the-box macros that work with it: you need to tell the compiler to make you some macros (with a macro ofc).

What makes 'em really nice is that they don't pollute the global namespace like structs. If you define a struct, no other library can define a different struct with the same name (same rules as modules, and that's the reason for this PR: the conflict betwen exprotobuf & protobuf both exporting `Protobuf` :facepalm:)

Here's what a read response might look like:

```elixir
iex(1)> Spear.stream!(conn, "es_supported_clients", raw?: true) |> Enum.take(-1) |> List.first
{:"event_store.client.streams.ReadResp", {:event,
 {:"event_store.client.streams.ReadResp.ReadEvent",
  {:"event_store.client.streams.ReadResp.ReadEvent.RecordedEvent",
   {:"event_store.client.shared.UUID",
    {:string, "06fc2d85-d9b8-438e-9bb1-c65bb8217674"}},
   {:"event_store.client.shared.StreamIdentifier", "es_supported_clients"}, 0,
   18446744073709551615, 18446744073709551615,
   [
     {"content-type", "application/json"},
     {"type", "grpc-client"},
     {"created", "16181816160401932"}
   ], "",
   "{\"languages\":[\"typescript\",\"javascript\"],\"runtime\":\"NodeJS\"}"},
  :undefined, {:no_position, {:"event_store.client.shared.Empty"}}}}}
```

So it's not very pretty. The basic definition of a record is

> a tuple where the first element is an atom

so yep, ok- and error-tuples are records :thinking:

but it's really annoying to work with records raw because the fields are done
positionally (ordered):

```elixir
{_read_resp, _content = {:event, {_read_resp_read_event, _event, _link, _position}}
```

and the names are not included. So it's a bit confusing to look at and work with because you need things in the right order but you can't immediately see what name each field has. Enter the `Record` built-in module in Elixir! (Also see `Spear.Records` in this PR). You can use `Record.defrecord/2` with `Record.extract/2` or `Record.extract_all/1` to define a record from an `.hrl` file like those that `:gpb` generates from the `.proto` files and it will define a set of macros to interact with the record and produce records in that shape.

E.g. for `:"event_store.client.streams.ReadResp"` from the above example:

```elixir
defmodule RecordDemo do
  require Record

  Record.defrecord(
    _custom_name = :read_resp,
    _record_name_as_in_hrl = :"event_store.client.streams.ReadResp",
    _record_definition_as_keyword_list = Record.extract(:"event_store.client.streams.ReadResp", from: "src/spear_proto_streams.hrl")
  )
end

iex> require RecordDemo
iex> event = Spear.stream!(conn, "es_supported_clients", raw?: true) |> Enum.take(-1) |> List.first
# that big ugly tuple you saw above
iex> match?(RecordDemo.read_resp(), event)
true
iex> RecordDemo.read_resp(content: content) = event
# that big ugly tuple you saw above
iex> content
{:event,
 {:"event_store.client.streams.ReadResp.ReadEvent",
  {:"event_store.client.streams.ReadResp.ReadEvent.RecordedEvent",
   {:"event_store.client.shared.UUID",
    {:string, "06fc2d85-d9b8-438e-9bb1-c65bb8217674"}},
   {:"event_store.client.shared.StreamIdentifier", "es_supported_clients"}, 0,
   18446744073709551615, 18446744073709551615,
   [
     {"content-type", "application/json"},
     {"type", "grpc-client"},
     {"created", "16181816160401932"}
   ], "",
   "{\"languages\":[\"typescript\",\"javascript\"],\"runtime\":\"NodeJS\"}"},
  :undefined, {:no_position, {:"event_store.client.shared.Empty"}}}}
```

And you can use the `read_resp/0`/`read_resp/1`/`read_resp/2` macros generated by `Record.defrecord/2` to generate those messages too. (See some of the stuff in `Spear.Writing` and `Spear.Reading` in this PR for that).

### CODE REVIEW

File: lib/spear/connection/request.ex [hunk 1]
```diff
@@ -216,12 +222,12 @@ defmodule Spear.Connection.Request do
 
   def handle_data(%__MODULE__{type: {:on_data, on_data}} = request, new_data) do
     case Spear.Grpc.decode_next_message(request.response.data <> new_data, request.response.type) do
-      {%ReadResp{content: {:confirmation, %{subscription_id: _}}}, rest} ->
+      {read_resp(content: {:confirmation, _confirmation}), rest} ->
```
Comment (the-mikedavis) [inline lib/spear/connection/request.ex:225]: as you can see here, the `Record.defrecord/2` produced macros made it really easy to convert from the structs of `elixir-protobuf/protobuf` to gpb records

File: priv/protos/README.md [hunk 2]
```diff
@@ -8,19 +8,16 @@ EventStore/EventStore@6eedacecd6a4da8ab705a82d5229f5c630f60277
 
 ## How to generate the Elixir modules from these protobufs
```
Comment (the-mikedavis) [inline priv/protos/README.md:9]: this file is a good read on how we generate the erl/hrl files from the protos

**Review by bjornrud — APPROVED** [review]
`:ok`

poodle intently listening and note taking

File: priv/protos/README.md (hunk 2 above)
Comment (the-mikedavis) [reply priv/protos/README.md:9]: oh crap just realized it says Elixir :facepalm: 

```suggestion
## How to generate the erl/hrl files from these protobufs
```

github is my ide

**Review by tonyvanriet — APPROVED** [review]
:muscle:

**Pull request merged**